lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -358,6 +358,133 @@ struct ConvolutionToConv2dPattern
+    auto outputType = mlir::cast<RankedTensorType>(op.getResult().getType());
+    auto shape = outputType.getShape();
+
+    // start indices of the gather op:
+    auto startIndices = op.getStartIndices();
+    auto startIndicesType =
+        mlir::cast<RankedTensorType>(startIndices.getType());
+
+    // slice sizes of the gather op
+    auto sliceSizes = op.getSliceSizes();
+    auto offsetDims = op.getOffsetDims();
+    // collapsed slice dims of the gather op
+    auto collapsedSliceDims = op.getCollapsedSliceDims();
+
+    if (shape.size() > 1) {
+      auto hiddenDim = shape[shape.size() - 1];
+      // check if sliceSizes has more than one element
+      if (sliceSizes.size() <= 1) {
+        return rewriter.notifyMatchFailure(op, "Did not satisfy sliceSizes");
+      }
+      // check if sliceSizes is [1, hiddenDim]
+      if (sliceSizes[0] != 1 || sliceSizes[1] != hiddenDim) {
+        return rewriter.notifyMatchFailure(op, "Did not satisfy sliceSizes");
+      }
+    }
+
+    // check if offsetDims is [2]
+    if (offsetDims.size() > 1 || offsetDims[0] != 2) {
+      return rewriter.notifyMatchFailure(op, "Did not satisfy offsetDims");
+    }
+
+    // check if collapsedSliceDims is [0]
+    if (collapsedSliceDims.size() > 1 || collapsedSliceDims[0] != 0) {
+      return rewriter.notifyMatchFailure(op,
+                                         "Did not satisfy collapsedSliceDims");
+    }
+
+    // check if startIndices and output have same shape, if not, check if
+    // reshape is possible can reshape startIndices to remove the last dimension
+    // if it is 1
+    if (shape.size() == startIndicesType.getShape().size() &&
+        startIndicesType.getShape()[shape.size() - 1] != 1) {
+      return rewriter.notifyMatchFailure(op,
+                                         "Did not satisfy startIndicesType");
+    }
+
+    return success();
+  }
+
+  ttir::ReshapeOp createReshapeOp(PatternRewriter &rewriter, Location loc,
+                                  Value input,
+                                  ::llvm::ArrayRef<int64_t> shapei64,
+                                  ::mlir::ArrayAttr operandConstraints) const {
+
+    // reshape start indices (input) to remove the last dimension
+    auto ty = mlir::cast<RankedTensorType>(input.getType());
+    auto output = rewriter.create<tensor::EmptyOp>(
+        loc, llvm::ArrayRef<int64_t>(shapei64), ty.getElementType());
+    std::vector<int32_t> shapei32(shapei64.begin(), shapei64.end());
+    auto shape_attr = rewriter.getI32ArrayAttr(shapei32);
+
+    return rewriter.create<ttir::ReshapeOp>(
+        loc, mlir::RankedTensorType::get(shapei64, ty.getElementType()), input,
+        output, shape_attr, operandConstraints);
+  }
+
+  LogicalResult
+  matchAndRewrite(ttir::GatherOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    // check if legal to convert to embedding
+    LogicalResult err = checkBasicLegality(op, rewriter);
```

Comment:
Did we consider fall backing to TTNN SliceOp instead of EmbeddingOp?

Reply:
StableHLOs gather is a very generic op, however, in all of the cases we've seen thus far, it's used to implement a simple embedding. While it may be possible to decompose it into slice (more likely multiple slice ops and a concat) this would unnecessarily complicate the conversion and resulting graph.  

When we come across instances of gather that are not embedding, we can expand this to add conversion to slice ops.

Reply:
I see. In my opinion it would make sense to decompose it into slice ops and a concat. However, if this is a common case seen for the embedding in StableHLO, in other words, if there is no other way to get to the TTNN embedding op through StableHLO, then I am fine with it. 
With this in mind, I'd add comment in the decomposition pattern rewriter, that explicitly states why we are doing this conversion. Right now it's not clear from the code why we chose this solution.

cc @ddilbazTT

Reply:
torch.embeddings is lowered into stablehlo.gather through torch-mlir, there may be other FEs that lower embeddings differently, but I can't really see anything else in the stablehlo op set that would fit. I suppose you could use a (sparse) matmul, but then we're back in buda land 😄 

This is a very generalized op (similar to reduce) that will likely need many lowering paths to be fully supported. This PR addresses the most common use case we've seen and we can add other paths as we find the need for them. We could add a comment to that regard, but I'm not sure how much additional value that would provide, i.e `this is an embedding lookup so we're lowering it to an embedding lookup`.

Reply:
I see. 
> but I'm not sure how much additional value that would provide, i.e this is an embedding lookup so we're lowering it to an embedding lookup.

I just suggested to have a more insightful comment, why specifically GatherOp is lowered to an EmbeddingOp. That's not clear to someone new to the compiler, unless they know details about `torch.embedding -> stablehlo.gather` conversion.